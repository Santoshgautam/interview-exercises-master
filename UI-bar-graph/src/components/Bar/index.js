import React from "react";

const Bar = ({ data, label, max }) => {
  return (
    <div className="bar" style={{ height: `${(data / max) * 100}%` }}>
      <span className="bar-value">{data}</span>
      <span className="bar-label">{label}</span>
    </div>
  );
};

export default Bar;
