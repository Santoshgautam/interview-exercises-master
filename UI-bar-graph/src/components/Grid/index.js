import React from "react";

const Grid = ({ maxStep, step, start }) => {
  const grids = [];
  let value = maxStep;

  while (value >= start) {
    grids.push(
      <div className="grid-y" key={value}>
        <div className="grid-label">{value}</div>
        <div className="grid-line"></div>
      </div>
    );
    value -= step;
  }
  return (
    <div className="axis-y">
      <div className="axis"></div>
      {grids}
    </div>
  );
};

export default Grid;
