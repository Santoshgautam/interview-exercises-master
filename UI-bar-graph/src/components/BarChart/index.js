import React from "react";
import Bar from "../Bar";
import Grid from "../Grid";

const BarChart = ({ data = [], labels = [] }) => {
  const maxStep = Math.ceil(data.reduce((r, d) => (r < d ? d : r), 0) / 10) * 10;

  return (
    <div className="bar-chart">
      <Grid maxStep={maxStep} step={10} start={0} />
      <div className="bar-chart__bars">
        {data.map((d, i) => (
          <Bar key={i} data={d} label={labels[i]} max={maxStep} />
        ))}
      </div>
    </div>
  );
};

export default BarChart;
