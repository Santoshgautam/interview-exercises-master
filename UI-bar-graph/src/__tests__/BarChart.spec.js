import BarChart from "../components/BarChart";
import Bar from "../components/Bar";
import Grid from "../components/Grid";

describe("BarChart Component", () => {
  let wrapper, consoleError;

  beforeEach(() => {
    wrapper = shallow(<BarChart />);
    consoleError = console.error;
    console.error = jest.spyOn(console, "error");
  });

  afterEach(() => {
    console.error = consoleError;
  });

  it("renders without crashing", () => {
    expect(wrapper.exists()).toBe(true);
  });

  it("matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("displays correct number of Bar components", () => {
    const sample = require("../constants/data.json");
    wrapper = shallow(<BarChart data={sample.data} labels={sample.labels} />);
    expect(wrapper.find(Bar)).toHaveLength(sample.data.length);
  });

  it("displays a Grid component", () => {
    const sample = require("../constants/data.json");
    wrapper = shallow(<BarChart data={sample.data} labels={sample.labels} />);
    expect(wrapper.find(Grid)).toHaveLength(1);
  });
});
