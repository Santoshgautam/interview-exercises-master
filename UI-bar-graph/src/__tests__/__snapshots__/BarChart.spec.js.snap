// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarChart Component matches the snapshot 1`] = `ShallowWrapper {}`;
